frontend/src/App.res

```diff
@@ -69,43 +61,86 @@ let getTestMetrics = (item: GetBenchmarks.t_benchmarks): BenchmarkTest.testMetri
-  BeltHelpers.Array.findIndexRev(benchmarks, (item: GetBenchmarks.t_benchmarks) => {
-    item.pull_number == None && item.test_name == testName
-  })
-}
+module BenchmarkView = {
+  let decodeRunAt = runAt => runAt->Js.Json.decodeString->Belt.Option.map(Js.Date.fromString)
 
-module BenchmarkResults = {
-  @react.component
-  let make = (~benchmarks: array<GetBenchmarks.t_benchmarks>, ~synchronize, ~repo_id) => {
-    let data = benchmarks->Belt.Array.map(getTestMetrics)
-    let selectionByTestName =
-      data->Belt.Array.reduceWithIndex(Belt.Map.String.empty, BenchmarkTest.groupByTestName)
+  let decodeMetrics = metrics =>
+    metrics
+    ->Belt.Option.getExn
+    ->Js.Json.decodeObject
+    ->Belt.Option.getExn
+    ->jsDictToMap
+    ->Belt.Map.String.map(v => BenchmarkTest.decodeMetricValue(v))
 
-    let comparisonMetricsByTestName = {
-      Belt.Map.String.mapWithKey(selectionByTestName, (testName, _) => {
-        // TODO: Use the index load the data from master and add an annotation.
-        switch getLatestMasterIndex(~testName, benchmarks) {
-        | Some(idx) => Some(benchmarks[idx]->getTestMetrics)
-        | None => None
-        }
-      })
-    }
+  let isMainBranch = pullNumber => {
+    Belt.Option.isNone(pullNumber)
+  }
 
-    let graphs = {
-      selectionByTestName
-      ->Belt.Map.String.mapWithKey((testName, testSelection) => {
-        let comparisonMetrics = Belt.Map.String.getExn(comparisonMetricsByTestName, testName)
-        <BenchmarkTest
-          ?comparisonMetrics synchronize key={testName} data testName testSelection repo_id
-        />
+  let makeBenchmarkData = (benchmarks: array<GetBenchmarks.t_benchmarks>) => {
+    benchmarks->Belt.Array.reduce(BenchmarkData.empty, (acc, item) => {
+      item.metrics
+      ->decodeMetrics
+      ->Belt.Map.String.reduce(acc, (acc, metricName, value) => {
+        BenchmarkData.add(
+          acc,
+          ~testName=item.test_name,
+          ~metricName,
+          ~runAt=item.run_at->decodeRunAt->Belt.Option.getExn,
+          ~commit=item.commit,
+          ~value,
+        )
       })
-      ->Belt.Map.String.valuesToArray
+    })
+  }
+
+  // Urql uses records to represent query results.
+  // Records have nominal types which makes two records with the same fields incompatible.
+  // This function assumes that the main benchmarks and the comparison benchmarks have the same structural type.
+  let unsafeCastComparisonBenchmarks = (
+    comparison: array<GetBenchmarks.t_comparisonBenchmarks>,
+  ): array<GetBenchmarks.t_benchmarks> => Belt.Array.map(comparison, Obj.magic)
+
+  @react.component
+  let make = (~repoId, ~pullNumber=?, ~startDate, ~endDate) => {
+    let ({ReasonUrql.Hooks.response: response}, _) = {
+      let startDate = Js.Date.toISOString(startDate)->Js.Json.string
+      let endDate = Js.Date.toISOString(endDate)->Js.Json.string
+      let comparisonLimit = isMainBranch(pullNumber) ? 0 : 10
```

Comment:
Since it looks like a derived value from the `pullNumber`, wouldn't that be better to move it within the `makeGetBenchmarksVariables` itself?

Also, I understand that we don't want any comparison results when showing data for master, but what's the reason for limiting them when showing data for PR's (and why `10`)?

Reply:
Agreed on the first point.

Here's my intended behaviour:
* When requesting data for master, we don't want to fetch any comparison data (hence `limit 0`).
* When requesting data for a PR we want to fetch data from master.
  * The master data is only used to contextualise the PR data in the graphs, we don't need the entire master dataset.
  * 10 seems like a sufficient number of datapoints for this, but I'm happy to change it to something else. :)